src/log.js

```diff
@@ -240,79 +250,14 @@ class Log extends GSet {
     if (!isDefined(log)) throw LogError.LogNotDefinedError()
     if (!Log.isLog(log)) throw LogError.NotALogError()
 
-    // Verify the entries
-    // TODO: move to Entry
-    const verifyEntries = async (entries) => {
-      const isTrue = e => e === true
-      const getPubKey = e => e.getPublic ? e.getPublic('hex') : e
-      const checkAllKeys = (keys, entry) => {
-        const keyMatches = e => e === entry.key
-        return keys.find(keyMatches)
-      }
-      const pubkeys = this._keys.map(getPubKey)
-
-      const verify = async (entry) => {
-        if (!entry.key) throw new Error("Entry doesn't have a public key")
-        if (!entry.sig) throw new Error("Entry doesn't have a signature")
-
-        if (this._keys.length === 1 && this._keys[0] === this._key ) {
-          if (entry.id !== this.id) 
-            throw new Error("Entry doesn't belong in this log (wrong ID)")
-        }
-
-        if (this._keys.length > 0 
-            && !this._keys.includes('*') 
-            && !checkAllKeys(this._keys.concat([this._key]), entry)) {
-          console.warn("Warning: Input log contains entries that are not allowed in this log. Logs weren't joined.")
-          return false
-        }
-
-        try {
-          await Entry.verifyEntry(entry, this._keystore)
-        } catch (e) {
-          throw new Error(`Invalid signature in entry '${entry.hash}'`)
-        }
-
-        return true
-      }
-
-      const checked = await pMap(entries, verify)
-      return checked.every(isTrue)
-    }
-
-    const difference = (log, exclude) => {
-      let stack = Object.keys(log._headsIndex)
-      let traversed = {}
-      let res = {}
-
-      const pushToStack = hash => {
-        if (!traversed[hash] && !exclude.get(hash)) {
-          stack.push(hash)
-          traversed[hash] = true
-        }
-      }
+    // Get the difference of the logs
+    const newItems = Log.difference(log, this)
 
-      while (stack.length > 0) {
-        const hash = stack.shift()
-        const entry = log.get(hash)
-          if (entry && !exclude.get(hash) && entry.id === this.id) {
-          res[entry.hash] = entry
-          traversed[entry.hash] = true
-          entry.next.forEach(pushToStack)
-        }
-      }
-      return res
-    }
-
-    // Merge the entries
-    const newItems = difference(log, this)
-
-    // if a key was given, verify the entries from the incoming log
-    if (this._key && this._key.getPublic) {
-      const canJoin = await verifyEntries(Object.values(newItems))
-      // Return early if any of the given entries didn't verify
-      if (!canJoin)
-        return this
+    // If we have an entry validator, verify that all new entries can be joined
+    // with this log, throws an error if fails
+    if (this._entryValidator) {
+      const verify = async (entry) => await Entry.verify(entry, this._entryValidator)
```

Comment:
If I read the changes in the PR correctly, this would validate that the *given key* was able to write to the log. And we give the key per-log, ie. for one user at a time. This has the same effect as described in the [comment above](https://github.com/orbitdb/ipfs-log/pull/131/files#r202507766)

Reply:
This shouldn't check for the current key, that key is only provided to the Log constructor so it could still be used as a fallback to the `clockId`. The key that'll be used to sign entries is available at the access controller so `ipfs-log` doesn't know which key is used to sign the entry, only that it has a signing function to call that will provide it.